[Training Manual] 2/3 Chapter 7 (#3110)
qgis/QGIS-Documentation: source/docs/training_manual/vector_analysis/network_analysis.rst

===============================================================================

Calculating the shortest distance between two points is a commonly cited use
for GIS. This tool can be found in the :guilabel:`Processing` toolbox.

**The goal for this lesson:** learn to use :guilabel:`Network analysis`
algorithms.

|basic| |FA| The Tool and the Data
-------------------------------------------------------------------------------

You can find all the network analysis algorithms in the
:menuselection:`Processing --> Network Analysis` menu. You can see that there
are many tools available:

.. image:: img/select_road_graph_plugin.png
   :align: center

Open the project :file:`exercise_data/network_analysis/network.qgz`, it contains
two layers:

1. :guilabel:`network_point`
2. :guilabel:`network_lines`

As you can see the :guilabel:`network_lines` layer has already a style that helps
to understand the road network.

> |basic| |FA| Use the Tool

Author: +1 to remove |FA|, -1 to remove the section. It cannot be straightforward to understand what the alg suite makes, so a brief description IMHO can help a lot

-------------------------------------------------------------------------------

Depending on the aim of your analysis you may want to calculate the shortest or
the fastest path between two point of a network.

The tool has many different options: e.g. it allows you to choose manually the
starting and ending point on the map or you can define start/end point on the
map while starting/ending points are taken from the point layer.

Let's start.

Shortest path (point to point)
------------------------------
The :menuselection:`Network analysis --> Shortest path (point to point)` allows
you to calculate the shortest distance between two manually selected points on
the map.

> In this example we will calculate the **shortest** (not fastest) path between two

Reviewer: @ghtmtt's suggestion is good - should not be changed.

> In this example we will calculate the **shortest** (not fastest) path between two

Reviewer 2: @havatv More details why it's good, please? There are two options in the tool: shortest and fastest (mentioned few lines above). So If I choose to go the shortest way, then it's obvious that I do not choose fastest. But what does selecting shortest mean? No idea! Is it shortest by the road, shortest at flying direction like a buffer...? I think there's a(more?) plus-value in explaining the kind of shortest we are going to explore than mentioning that shortest is not fastest.

> In this example we will calculate the **shortest** (not fastest) path between two

Reviewer: Sorry I was not precise enough.  Removing "(not fastest)" is probably OK, but not adding "(following the network, not the flying distance)"

> In this example we will calculate the **shortest** (not fastest) path between two

Author: I think that we are putting to much entropy in this. The user has just to understand that the calculation is the shortest and not the fastest

> In this example we will calculate the **shortest** (not fastest) path between two

Reviewer: Can we agree on:
`In this example we will calculate the **shortest** path between two points in the network.`
Should not be necessary to say that one needs to follow the lines / edges of the network - it is implicit when we say path.

points.

In the following image we choose these two points as starting and ending point

> for the analysis, feel free to change them and test other locations.

Author: Ok section moved. I don't get what you mean by `isn't it only a two features layer`?

.. image:: img/start_end_point.png
   :align: center

* Open the :guilabel:`Shortest path (point to point)` algorithm;
* Select :guilabel:`network_lines` for :guilabel:`Vector layer representing network`;
* Let ``Shortest`` in the :guilabel:`Path type to calculate` parameter;
* Click on the |browseButton| button next to the :guilabel:`Start Point (x,y)`

>   and choose the location you want. The menu is filled with the coordinates of

Author: correct, changed. 

> IOW is the point layer needed here?

nope. It is just needed to show the location to the user so that the result will be (more or less) the same

  the clicked point;
* Do the same thing for :guilabel:`End Point(x,y)`;
* Click on the :guilabel:`Run` button:

.. image:: img/shortest_point.png
   :align: center

A new line layer is created representing the shortest path between the chosen
points. Uncheck the :guilabel:`network_lines` layer to see the result better:

.. image:: img/shortest_point_result.png
   :align: center

Let's open the attribute table of the output layer. It contains three fields,

> representing the coordinates of the staring and ending points and the

Reviewer: ```suggestion
representing the coordinates of the starting and ending points and the
```

**cost**.

Depending on what you choose in the :guilabel:`Path type to calculate` it will
contain either the distance, in layer units, between points (``Shortest``) or
the time, in hour units (``Fastest``).

> In out case, the *shortest* distance between the chosen points is around ``1000``

Reviewer: ```suggestion
In our case, the *shortest* distance between the chosen points is around ``1000``
```

meters:

.. image:: img/shortest_point_attributes.png
   :align: center

.. _backlink-network_analysis_1:

|moderate| |TY| Fastest path
--------------------------------------------------------------------------------
Redo the previous exercise, but change the :guilabel:`Path type to calculate`

> from ``Shortest`` to ``Fastest``.

Author: good idea! but now the exercise become moderate ;)

> How much time do you need to go from the start to the end point?

Author: I think that with the rephrased section this question fits good

|moderate| |FA| Advanced options
-------------------------------------------------------------------------------

Let's explore some more options of the Network Analysis tools. In the :ref:`previous
exercise <backlink-network_analysis_1>` we calculated the **fastest** route
between two points. As you can imagine, the time depends on the travel **speed**.

We will use the same layers and same starting and ending points of the previous
exercises, but again, feel free to experiment with other options.

* Open the :guilabel:`Shortest path (point to point)` algorithm;
* Fill the :guilabel:`Input layer`, :guilabel:`Start Point (x,y)` and
  :guilabel:`End Point (x,y)` as we did before;
* Choose ``Fastest`` as the :guilabel:`Path type to calculate`;
* Open the :guilabel:`Advanced parameter` menu;
* Change the :guilabel:`Default speed (km/h)` from the default ``60`` value to
  ``4``;

  .. image:: img/shortest_path_advanced.png
     :align: center

* Click on :guilabel:`Run`.

Once the algorithm is finished, close the dialog and open the attribute table of
the output layer.

The :guilabel:`cost` field contains the value according to the speed parameter
you have chosen.

> We can convert the :guilabel:`cost` field from hours with fractions to the more

Author: > From here, sorry but i find the exercise less interesting: learning conversion from hours to minutes is not that fun.

it is not fun but I can bet on how many people could be confused on time values like 1.86 hours. 

> Giving a look to the osm data we used from the beginning, there's a maxspeed field which contains speed data. What about using it instead and more globally why not rely on the osm data we provide from the beginning?:

OSM data (especially with numbers like speed values) are not always reliable. Sometimes values are read as strings (because users add something like `10 kh` as the values causing all the column recognized as string). 

I thought to download data from OSM, but it could be 80% of the section explaining how to clean and prepare the data and 20% focusing on the exercise. I really want that users know what these algorithm make and not put efforts in cleaning the data.

readable *minutes* values.

* Open the field calculator by clicking on the |calculateField| icon and add the
  new field :guilabel:`minutes` by multiplying the :guilabel:`cost` field by 60:

  .. image:: img/shortest_path_conversion.png
     :align: center

That's it! Now you know how many minutes it will take to get from one point to
the other one.

|moderate| Service area (from layer)
-------------------------------------------------------------------------------
The :menuselection:`Network Analysis --> Service area (from layer)` algorithm

> can answers the question: given a point layer, what are all the reachable areas

Reviewer: ```suggestion
can answer the question: given a point layer, what are all the reachable areas
```

given a distance or a time value?

> .. note:: The :menuselection:`Network Analysis --> Service area (from points)`

Reviewer: ```suggestion
.. note:: The :menuselection:`Network Analysis --> Service area (from point)`
```

>     is the same algorithm but but it allows you to manually choose the point on

Reviewer: ```suggestion
    is the same algorithm, but it allows you to manually choose the point on
```

>     is the same algorithm but, it allows you to manually choose the point on the

Reviewer: ```suggestion
    is the same algorithm but, it allows you to manually choose the points on the
```

Reviewer: Since that name is "Service area (from points)", I guess that more than one point may be "chosen".

Reviewer 2: actually there's a typo in the alg name; it's **from point** and a single one can be set.

    the map.

Given a distance of ``250`` meters we want to know how far we can get from each
of the two points of the :guilabel:`network_point` layer.

* Uncheck all the layers except :guilabel:`network_point`;
* Open the :menuselection:`Network Analysis --> Service area (from layer)` algorithm;
* Choose :guilabel:`network_lines` for :guilabel:`Vector layer representing network`;
* Choose :guilabel:`network_points` for :guilabel:`Vector layer with start points`;
* Choose ``Shortest`` in :guilabel:`Path type to calculate`;
* Enter ``250`` in the :guilabel:`Travel cost` parameter;
* Click on :guilabel:`Run` and then close the dialog.

.. image:: img/service_area.png
   :align: center

The output layer represents the maximum path you can reach from the point features
given a distance of 250 meters:

.. image:: img/service_area_result.png
   :align: center

Cool isn't it?

> |hard| Shortest map with speed limit

Author: OK done. Service area is now the last exercise

-------------------------------------------------------------------------------
The Network analysis toolbox has other interesting options. Looking at the
following map:

.. image:: img/speed_limit.png
   :align: center

we would like to know the **fastest** route considering the **speed limits** of
each road (the labels represent the speed limits in km/h). The shortest path without
considering speed limits would of course be the purple path. But in that road
the speed limit is 20 km/h, while in the green road you can go at 100 km/h!

As we did in the first exercise, we will use the
:menuselection:`Network analysis --> Shortest path (point to point)` and we will
manually choose the start and end points.

* Open the :menuselection:`Network analysis --> Shortest path (point to point)`
  algorithm;
* Select :guilabel:`network_lines` for the :guilabel:`Vector layer representing network`
  parameter;
* Choose ``Fastest`` as the :guilabel:`Path type to calculate`;
* Click on the |browseButton| button next to the :guilabel:`Start Point (x,y)`

>   and choose the location you want. The menu is filled with the coordinates of

Author: correct, changed following the same structure as the first exercise

  the clicked point;
* Do the same thing for :guilabel:`End Point(x,y)`;
* Open the :guilabel:`Advanced parameter` menu;
* Choose the ``speed`` field as the :guilabel:`Speed Field` parameter. With this
  option the algorithm will take into account the speed values for each road;

> 

Author: That's too generic. I refactor the explanation with a new section so putting this *old* sentence back means to re-re factor the section again

  .. image:: img/speed_limit_parameters.png
     :align: center

* Click on the :guilabel:`Run` button:

Turn off the :guilabel:`network_lines` layer to better see the result.

.. image:: img/speed_limit_result.png
   :align: center

As you can see the fastest route does not correspond to the shortest one.

|IC|
-------------------------------------------------------------------------------

> Now you know how to use the :guilabel:`Network analysis` algorithm to solve

Reviewer: ```suggestion
Now you know how to use :guilabel:`Network analysis` algorithms to solve
```

===============================================================================

Calculating the shortest distance between two points is a commonly cited use

> for GIS. QGIS ships with this tool within the :guilabel:`Processing` toolbox.

Reviewer: This tool can be found in the :guilabel:`Processing` toolbox.

**The goal for this lesson:** learn to use the :guilabel:`Network analysis`
algorithms.

|basic| |FA| The Tool and the Data
-------------------------------------------------------------------------------

You can find all the network analysis algorithms in the
:menuselection:`Processing --> Network Analysis` menu. You can see that there
are many tool available, depending on which is the analysis you want to perform.

.. image:: img/select_road_graph_plugin.png
   :align: center

Open the project :file:`exercise_data/network_analysis/network.qgz`, it contains
two layers:

1. :guilabel:`network_point`
2. :guilabel:`network_lines`

As you can see the :guilabel:`network_lines` layer has already a style that helps
to understand the road network.

|basic| |FA| Use the Tool
-------------------------------------------------------------------------------

Depending on the aim of your analysis you may want to calculate the shortest or
the fastest path between two point of a network.

The tools has many different options: e.g. it lets allows you to choose manually
the starting and ending point on the map, or it lets you to use a point layer that

> contains starting and ending points.

Author: you have 2 chances:

1. many starting points defined in the layer and manually choose the end point
2. manually set the start point and end points are taken from the layer

  and choose the location you want. The menu is filled with the coordinates of
  the clicked point.
* Do the same thing for :guilabel:`End Point(x,y)`.
* Click on the **[Run]** button:

.. image:: img/shortest_point.png
   :align: center

A new line layer is created representing the shortest path between the chosen
points. Uncheck the :guilabel:`network_lines` layer to see better only the final
result:

.. image:: img/shortest_point_result.png
   :align: center

Let's open now the attribute table of the output layer. It is made of three
field, containing the coordinates of the staring and ending points and the
**cost** one.

Depending on what you choose in the :guilabel:`Path type to calculate` it will
contain either the distance, in layer units, between points (``Shortest``) or
the time, in hour units (``Fastest``).

In out case, the *shortest* distance between the chosen points is around ``1000``
meters:

.. image:: img/shortest_point_attributes.png
   :align: center

.. _backlink-network_analysis_1:

|moderate| |TY| Fastest path
--------------------------------------------------------------------------------

> Redo the previous exercise but change the :guilabel:`Path type to calculate`

Reviewer: comma between exercise and but

from ``Shortest`` to ``Fastest``.

> How much time do you need to go from the star to the end point?

Reviewer: star -> start

:ref:`Check your results <network-analysis-1>`

|moderate| |FA| Advanced options
-------------------------------------------------------------------------------

> Let's explore some more options of the Network Analysis tool. In the :ref:`previous

Reviewer: tool -> tools

exercise <backlink-network_analysis_1>` we calculated the **fastest** route

> between two points. As you can imagine, the time changes accordingly to the

Reviewer: the time depends on the travel **speed**.

travel **speed**.

We will use the same layers and same starting and ending points of the previous

> exercises, but again, fell free to experiment other options.

Reviewer: fell -> feel

* Open the :guilabel:`Shortest path (point to point)` algorithm.
* Fill the :guilabel:`Input layer`, :guilabel:`Start Point (x,y)` and
  :guilabel:`End Point (x,y)` as we did before.
* Choose ``Fastest`` as the :guilabel:`Path type to calculate`.
* Open the :guilabel:`Advanced parameter` menu.
* Change the :guilabel:`Default speed (km/h)` from the default ``60`` value to
  ``4``.

  .. image:: img/shortest_path_advanced.png
     :align: center

* Click on **[Run]**.

Once the algorithm is finished, close the dialog and open the attribute table of
the output layer.

The :guilabel:`cost` field contains the value according to the speed parameter
you have chosen.

> We can convert the :guilabel:`cost` field from hour fraction to the more readable

Reviewer: "hour fractions"  or "hours with fractions"

*minutes* values.

* Open the field calculator by clicking on the |calculateField| icon and add the

>   new field :guilabel:`minutes` by multiplying the :guilabel:`cost` field for 60:

Reviewer: field by 60

  .. image:: img/shortest_path_conversion.png
     :align: center

That's it! Now you know how many minutes it will take to get from one point to
the other one.

|moderate| Service area (from layer)
-------------------------------------------------------------------------------
The :menuselection:`Network Analysis --> Service area (from layer)` algorithm

> answers to the question: given a point layer what are all the reachable areas

Reviewer: answers to -> can answer

given a distance or a time value?

.. note:: The :menuselection:`Network Analysis --> Service area (from points)`

>     is the same algorithm but you can manually choose the point on the map.

Reviewer: but it allows you to manually

Given a distance of ``250`` meters we want to know how far we can get from each
of the two points of the :guilabel:`network_point` layer.

* Uncheck all the layers except :guilabel:`network_point`.
* Open the :menuselection:`Network Analysis --> Service area (from layer)` algorithm.
* Choose :guilabel:`network_lines` for :guilabel:`Vector layer representing network`.
* Choose :guilabel:`network_points` for :guilabel:`Vector layer with start points`.
* Choose ``Shortest`` in :guilabel:`Path type to calculate`.
* Enter ``250`` in the :guilabel:`Travel cost` parameter.
* Click on **[Run]** and then close the dialog.

.. image:: img/service_area.png
   :align: center

> The output layer represent the maximum path you can reach from the point features

Reviewer: represents (plural)

given a distance of 250 meters:

.. image:: img/service_area_result.png
   :align: center

Cool isn't it?

|hard| Shortest map with speed limit
-------------------------------------------------------------------------------
The Network analysis toolbox has other interesting options. Looking at the
following map:

.. image:: img/speed_limit.png
   :align: center

> we would know the **fastest** route considering the **speed limits** of each

Reviewer: we would like to know

considering speed limits would of course be the purple path. But in that road
the speed limit is 20 km/h, while in the green road you can go at 100 km/h!

As we did in the first exercise, we will use the
:menuselection:`Network analysis --> Shortest path (point to point)` and we will
manually choose the start and end points.

* Open the :menuselection:`Network analysis --> Shortest path (point to point)`
  algorithm.
* Select :guilabel:`network_lines` for the :guilabel:`Vector layer representing network`
  parameter.
* Choose ``Fastest`` as the :guilabel:`Path type to calculate`.
* Click on the |browseButton| button next to the :guilabel:`Start Point (x,y)`
  and choose the location you want. The menu is filled with the coordinates of
  the clicked point.
* Do the same thing for :guilabel:`End Point(x,y)`.
* Open the :guilabel:`Advanced parameter` menu.
* Choose the ``speed`` field as the :guilabel:`Speed Field` parameter. With this
  option the algorithm will take into account the speed values for each road.

  .. image:: img/speed_limit_parameters.png
     :align: center

* Click on the **[Run]** button:

> Turn off the :guilabel:`network_lines` layer to see better the final result.

Reviewer: to better see the result.

===============================================================================

Calculating the shortest distance between two points is a commonly cited use
for GIS. This tool can be found in the :guilabel:`Processing` toolbox.

**The goal for this lesson:** learn to use :guilabel:`Network analysis`
algorithms.

|basic| |FA| The Tool and the Data
-------------------------------------------------------------------------------

You can find all the network analysis algorithms in the
:menuselection:`Processing --> Network Analysis` menu. You can see that there
are many tools available:

.. image:: img/select_network_algorithms.png
   :align: center

Open the project :file:`exercise_data/network_analysis/network.qgz`, it contains
two layers:

1. :guilabel:`network_points`
2. :guilabel:`network_lines`

As you can see the :guilabel:`network_lines` layer has already a style that helps
to understand the road network.

.. image:: img/network_map.png
   :align: center

|basic| |FA| Use the Tool
-------------------------------------------------------------------------------

> The tool provides ways to calculate either the shortest or the fastest path

Reviewer: What about "The shortest path tools provide ..."

between two points of a network, given:

* starting and ending points selected on the map;

> * starting or ending points taken from a point layer.

Reviewer: Why not just list the three possibilities:
```
* start point and end point selected on the map;
* start point selected on the map and end points taken from a point layer;
* start points taken from a point layer and end point selected on the map;
```
